Add Solving problems with custom extensions blog post (#964)
quarkusio/quarkusio.github.io: _posts/2021-04-19-solving-problems-with-custom-extensions.adoc

---
layout: post
title: 'Solving problems with custom Quarkus extensions'
date: 2021-04-19
tags: development-tips extensions
synopsis: If fast startup and live reload weren't enough for you, maybe custom extensions will convince you that you can achieve great things with Quarkus.
author: gsmet
---

From time to time, I see tweets or articles claiming they don't see the point of Quarkus because "who needs fast startup?", "I have plenty of memory" or "what is the point of live reload?".

I could write an article to debunk these arguments and explain how the latter makes your development workflow much more efficient and how the former makes the latter possible, even if fast boots are not your thing.

> But for the sake of this blog post, let's admit these persons are absolutely right and they are not good reasons to use Quarkus.

Reviewer: ```suggestion
But for the sake of this blog post, let's admit these persons are absolutely right and there are no good reasons to use Quarkus.
```

Author: I used `these` instead of `they`. Your proposal changes significantly the meaning.

Reviewer: Makes sense :)

So now what? Back to `<insert your favorite framework here>`? Not so fast.

> Quarkus didn't achieve fast startup and low memory footprint by using dark magic or lazy loading tricks but by rethinking entirely the way Java applications are bootstrapped.

Reviewer 2: ```suggestion
Quarkus didn't achieve fast startup and low memory footprint by using dark magic or lazy loading tricks but by entirely rethinking the way Java applications are bootstrapped.
```

> Quarkus didn't achieve fast startup and low memory footprint by using dark magic or lazy loading tricks but by rethinking entirely the way Java applications are bootstrapped.

Reviewer: I kinda feel the blog could just start here on line 18 - nothing technical wrong per say on the lines above but they give a bit of a negative vibe.

> The whole point of Quarkus is to move as much work as possible to the build and this journey made us build a framework to push work at build time and this framework can be leveraged in Quarkus extensions.

Reviewer 2: ```suggestion
The whole point of Quarkus is to move as much work as possible to the build-time and this journey made us create a framework to push work at build time and this framework can be leveraged in Quarkus extensions.
```

> A Quarkus extension? That sounds like a lot of work?

No, really, it is not.

> You can develop your own extensions very easily and they can solve some unordinary issues in a very simple way.

Reviewer 2: ```suggestion
You can develop your own extensions very easily and they can solve some unique issues in a very simple way.
```

Author: I used `out of the ordinary` because that's the meaning I want to convey.

----

It will ask for a `groupId` - let's keep the default `org.acme` - and an extension id - I went for `message-transformers-as-beans`.

Then you can import your new extension into your favorite IDE.

== Structure of the extension

There is a lot to say about extensions but, in the context of this blog post, we will keep it short.
The extension is composed of three Maven modules:

* The parent module - nothing to see here
* The deployment module - this is the one of interest for our blog post
* The runtime module - in this blog post, we won't modify it

Let's keep it simple: the deployment module is what will be used at build time, the runtime module is what will be used at runtime.

In our case, we want to declare new beans and this is something we do at build time, so deployment module, here we come!

== Processors and build steps

If you have a look at your `deployment` module, you will see a `MessageTransformersAsBeansProcessor` and you can see a method annotated with the `@BuildStep` annotation in it.

Quarkus build is populated by these build steps and they are following a consumer/producer model with dependency injection.

> The items being consumed and produced are called ``BuildItem``s.

Reviewer 2: ```suggestion
The items being consumed and produced are called `BuildItem`s.
```

Author: That's how you're supposed to do things when you have a character just after the backtick.

Reviewer 2: But what add the backtick in the first place?

> The preexisting build step is easy to understand.

Reviewer 2: ```suggestion
The build step that is automatically generated is easy to understand.
```

It produces a `FeatureBuildItem` which will be consumed by Quarkus startup and you will see the extension name in the list displayed by Quarkus at startup:

[source]
----

> INFO [io.quarkus] my-app 1.0-SNAPSHOT on JVM (powered by Quarkus 1.13.2.Final) started in 1.535s. Listening on: http://localhost:8080

Reviewer 2: This startup time seems pretty terrible TBH... Why is it so high?

Author: It's dev mode for another app, I just got the output from there :). I can make things a bit faster easily.

Reviewer 2: I'd say get some real numbers, as this doesn't look good :)

Author: Yeah I started an empty app and took these numbers.

INFO [io.quarkus] Profile dev activated. Live Coding activated.
INFO [io.quarkus] Installed features: [cdi, message-transformers-as-beans]
----

== The Jandex index

Now that we are done with scaffolding, let's think a bit about what we want to achieve:
we need to find all the classes in a given package whose name ends with `MessageTransformer`.

An important assumption of Quarkus is that the application lives in a closed world.

> You cannot dynamically add a jar to your Quarkus application and expect it to work.

Reviewer 2: ```suggestion
You cannot dynamically add a jar at runtime to your Quarkus application and expect it to work.
```

This can be achieved by adding another build step:

[source,java]
----
@BuildStep
void declareMessageTransformersAsBean(CombinedIndexBuildItem index, <1>
        BuildProducer<AdditionalBeanBuildItem> additionalBeans) { <2>
    List<String> messageTransformers = index.getIndex().getKnownClasses().stream() <3>
            .filter(ci -> !Modifier.isAbstract(ci.flags())) <4>
            .map(ci -> ci.name().toString()) <5>
            .filter(c -> c.startsWith("my.package.")) <6>
            .filter(c -> c.endsWith("MessageTransformer")) <7>
            .collect(Collectors.toList());

    additionalBeans.produce(new AdditionalBeanBuildItem.Builder() <8>
            .addBeanClasses(messageTransformers) 
            .setUnremovable() <9>
            .setDefaultScope(DotNames.APPLICATION_SCOPED) <10>
            .build());
}
----
<1> Consume the Jandex index
<2> Inject the additional beans producer
<3> Get all known classes from the index
<4> Filter out abstract classes
<5> Get the FQCN of the class
<6> Only keep classes from the root package we target
<7> Only keep ``MessageTransformer``s
<8> Produce an `AdditionalBeanBuildItem`

> <9> Make the beans unremovable

Reviewer 2: You might want to explain why this is done
